modules/fileout_netcdf/FONcArrayStructureField.cc

```diff
@@ -0,0 +1,366 @@
+    var_name = FONcUtils::gen_name(embed,d_varname, d_orig_varname);
+
+}
+
+/** @brief define the DAP array of structure field  in the netcdf file
+ *
+ * @param ncid The id of the NetCDF file
+ * @throws BESInternalError if there is a problem. 
+ */
+void
+FONcArrayStructureField::define( int ncid )
+{
+    BESDEBUG("fonc", "FONcArray::define() - defining array '" << d_varname << "'" << endl);
+
+    if (!d_defined) {
+
+        BESDEBUG("fonc", "FONcArray::define() - defining array of structure field: " << d_varname << "'" << endl);
+        auto i = struct_dims.begin();
+        auto e = struct_dims.end();
+        for (; i != e; i++) {
+            FONcDim *fd = *i;
+            fd->define_struct(ncid);
+            d_dim_ids.push_back(fd->dimid());
+            BESDEBUG("fonc", "FONcArray::define() - dim_id: " << fd->dimid() << " size:" << fd->size() << endl);
+        }
+
+        int stax = nc_def_var(ncid, var_name.c_str(), d_array_type, (int)(struct_dims.size()), d_dim_ids.data(), &d_varid);
+        if (stax != NC_NOERR) {
+            string err = (string) "fileout.netcdf - Failed to define variable " + d_varname;
+            FONcUtils::handle_error(stax, err, __FILE__, __LINE__);
+        }
+
+        stax = nc_def_var_fill(ncid, d_varid, NC_NOFILL, nullptr );
+        if (stax != NC_NOERR) {
+            string err = (string) "fileout.netcdf - " + "Failed to clear fill value for " + d_varname;
+            FONcUtils::handle_error(stax, err, __FILE__, __LINE__);
+        }
+    }
+}
+
+/** @brief Write the netcdf variable data out to the netcdf file
+ *
+ * Once the variable is defined, the value of the variable can be written out
+ *
+ * @param ncid The id of the netcdf file
+ * @throws BESInternalError if there is a problem writing the value
+ */
+void
+FONcArrayStructureField::write( int ncid )
+{
+    BESDEBUG( "fonc", "FONcArrayStructureField::write for var " << d_varname << endl ) ;
+
+    vector<char> data_buf;
+
+    // Obtain the total data buffer.
+    data_buf.resize(total_nelements*d_array_type_size);
+    char* data_buf_ptr = data_buf.data();
+
+    // Obtain the compound_buf, this is from gathering the data for individual fields.
+    vector<BaseType*> compound_buf = d_a->get_compound_buf();
+    for (unsigned i= 0; i<d_a->length_ll();i++) {
+        BaseType *cb = compound_buf[i];
+        if (cb->type()!=libdap::dods_structure_c){
+            throw BESInternalError("Fileout netcdf: This is not array of structure", __FILE__, __LINE__);
+        }
+        auto structure_elem = dynamic_cast<Structure *>(cb);
+        if (!structure_elem)
+            throw BESInternalError("Fileout netcdf: This is not array of structure", __FILE__, __LINE__);
+
+        //loop through the structure
+        Constructor::Vars_iter vi = structure_elem->var_begin();
+        Constructor::Vars_iter ve = structure_elem->var_end();
+
+        for (; vi != ve; vi++) {
```

Comment:
I think you can use structure_elem->variables() and the range-based for loop here.

Reply:
Yes, change the code according to your suggestions.